kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/testplugins/SaslPlainInitiationFilter.java

```diff
@@ -0,0 +1,155 @@
+        short handshakeMaxVersion = apiVersionsResponse.apiKeys().find(ApiKeys.SASL_HANDSHAKE.id).maxVersion();
+        short authenticateMaxVersion = apiVersionsResponse.apiKeys().find(ApiKeys.SASL_AUTHENTICATE.id).maxVersion();
+        LOGGER.info("Received API_VERSIONS response, maxHandshakeVersion: {}, maxAuthenticateVersion: {}", handshakeMaxVersion, authenticateMaxVersion);
+        this.saslHandshakeVersion = (short) Math.max(handshakeMaxVersion, 1);
+        this.saslAuthenticateVersion = (short) Math.max(authenticateMaxVersion, 2);
+        LOGGER.info("Will use handshakeVersion: {}, authenticateVersion: {}", this.saslHandshakeVersion, this.saslAuthenticateVersion);
+
+        if (state == State.AWAITING_API_VERSIONS) {
+            LOGGER.info("Sending SASL_HANDSHAKE request");
+            state = State.AWAITING_SASL_HANDSHAKE;
+            return context.<SaslHandshakeResponseData> sendRequest(
+                    requestHeader(ApiKeys.SASL_HANDSHAKE, saslHandshakeVersion),
+                    new SaslHandshakeRequestData()
+                            .setMechanism("PLAIN"))
+                    .thenCompose(handshakeResponse -> {
+                        if (handshakeResponse.errorCode() == Errors.NONE.code()) {
+                            LOGGER.info("Sending SASL_AUTHENTICATE request");
+                            state = State.AWAITING_SASL_AUTHENTICATE;
+                            return context.<SaslAuthenticateResponseData> sendRequest(
+                                    requestHeader(ApiKeys.SASL_AUTHENTICATE, saslAuthenticateVersion),
+                                    new SaslAuthenticateRequestData()
+                                            .setAuthBytes(("\0" + username + "\0" + password).getBytes(StandardCharsets.UTF_8)))
+                                    .thenCompose(authenticateResponse -> {
+                                        if (authenticateResponse.errorCode() == Errors.NONE.code()) {
+                                            for (BufferedRequest bufferedRequest : bufferedRequests) {
+                                                LOGGER.info("Forwarding buffered {} request", ApiKeys.forId(bufferedRequest.header.apiKey()));
+                                                context.forwardRequest(bufferedRequest.header, bufferedRequest.request);
+                                            }
+                                            state = State.FORWARDING;
+                                            LOGGER.info("Forwarding API_VERSIONS response");
+                                            return context.forwardResponse(header, apiVersionsResponse);
+                                        }
+                                        else {
+                                            throw new RuntimeException("authenticate failed: " + Errors.forCode(authenticateResponse.errorCode()).name());
```

Comment:
I'm confused, why are we throwing exceptions here rather than calling `context.clientSaslAuthenticationFailure` directly?